nn/transformer.py

```diff
@@ -0,0 +1,237 @@
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.activation = activation
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
+    self.dropout = dropout
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    """
+    Two possible forward variants of encoder, defined by self.norm_first
+    """
+    x = inp
+    if self.norm_first:
+      x = x + self._self_attention_block(nn.layer_norm(x, epsilon=self.norm_eps))
+      x = x + self._feed_forward_block(nn.layer_norm(x, epsilon=self.norm_eps))
+    else:
+      x = nn.layer_norm(x + self._self_attention_block(x), epsilon=self.norm_eps)
+      x = nn.layer_norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerEncoder(nn.Module):
+  """
+  Defines the full Encoder of the standard transformer
+  """
+  def __init__(self, encoder_layer: Union[TransformerEncoderLayer, Any], num_layers: int,
+               norm: Callable[[nn.LayerRef], nn.LayerRef] = None, layer_norm_eps: float = 1e-5):
+    """
+    :param encoder_layer: Encoder layer to be stacked num_layers times
+    :param num_layers: Number of layers
+    :param norm: normalization function
+    :param layer_norm_eps: Epsilon value for layer normalization
+    """
+    super().__init__()
+    import copy
+    self.layers = nn.ModuleList([copy.deepcopy(encoder_layer) for _ in range(num_layers)])
+
+    self.num_layers = num_layers
+    self.norm = norm
+    self.norm_eps = layer_norm_eps
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    """
+    Executes the encoder layer as often as in num_layers defined
+    """
+    output = inp
+    for mod in self.layers:
+      output = mod(output)
+
+    if self.norm is not None:
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output
+
+
+class TransformerDecoderLayer(nn.Module):
+  """
+  Defines one layer of a standard transformer decoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, layer_norm_eps: float = 1e-5,
+               norm_first: bool = False):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
```

Comment:
This is still wrong.

Reply:
would you change it for all Layers? Changed it for the Transformer Module, but thought hidden would be fine for Layers inside the Transformer. But ofc. can change every of those, if you prefer it that way.

Reply:
No, the documentation on a module should just be about this module isolated, no matter where or how it is used (i.e. it is not relevant whether it can be used such that it is hidden).